### PR TITLE
Draft KDD for multi unit item

### LINKED ISSUE #1089: Decision or KDD about multi unit item
Based on this comment: https://github.com/openmsupply/open-msupply/issues/973#issuecomment-1401208182, although it's in the context of pack to one conversion it's not strictly restricted to that area: 

> One key difficulty we have with existing mSupply is when stores are dealing with different ways of storing the same item. For example, a national warehouse will deal with bottles of Medicine A, whereas a health centre will want to deal with mLs of Medicine A. It's the same item, but different units. When we apply pack to one or transfers to this, it gets quite difficult - either the warehouse has to deal in mL (ie. one bottle is 1000mL so if they have 5 bottles, they write in 5000mLs), or health centres have to deal with very small fractions (.0001 of a bottle...). As you can imagine, this can get really messy and complicated.
> 
> It would be great if omSupply could deal with multiple units for an item

This is low priority placeholder issue to make sure we address it at some point.

### PR DESCRIPTION
**Pull request by andreievg**
Fixes #1089

This is draft KDD, please see issue for further material.

Markdown version: https://github.com/openmsupply/open-msupply/blob/42a4bbf4dbeaee1a84fe8fc60555d14f5e4d2655/decisions/multi-unit-item.md#L29

### CODE REVIEW

**Comment by CarlosNZ** [comment]
Yeah, looks good. Option 2 definitely seems like the best approach.


I just a few thoughts regarding this structure:

Screenshot 2023-06-15 at 2 28 12 PM

- Would be good to have a clear distinction between the "base" unit and all other units. I.e. every item _must_ have a base unit defined, and it should be the smallest non-divisible (i.e. always an integer) amount of the product (e.g. "tablet", "mL", etc.)
- All other units should be defined in terms of the base unit, not have references to each other. That way you can change/add/remove the higher-level units at a later time without having a cascading chain of changes to have to update, and you know there will *always* be a base unit, whereas other units could come and go.
- Presumably all internal calculation would always be done in terms of the "base" unit, and the other units only used for user-facing display?
- One other useful property for each different unit could be "isDivisible", to specify if that unit can have non-integer quantities. (e.g. "tablet" wouldn't, but "liters" might). Then again, enforcing that might be overly complicated if we're primarily dealing in base units internally, so maybe not.

If I was representing the above as a Type, it'd be something like:

```ts
interface item {
  name: string
  // ...other item properties
  baseUnit: {
    name: string
    description?: string // e.g. "in liquid form"
  }
  otherUnits?: Array<{
    name: string
    description?: string
    numberOfBaseUnits: number
    isDivisible: boolean // default true
  }>
}
```

**Review by mark-prins — COMMENTED** [review]
I think you're solving a different issue to the one that Adam is raising - though I'm not entirely sure what he is asking tbh. What I thought he wanted was a handling of doses and vials, and while the variants can be configured for that, you'd only see one or the other in a given facility. I think he'd like to be able to switch between either.

Would SOH, AMC and suggested quantities be displayed in the item variant which is selected for a given facility?

**Comment by andreievg** [comment]
Quick summary of our meeting:

We were not sure where original requirement came from and why/how we have so much focus on vaccines. @adamdewey the original customer to mention this, was it specifically in the context of vaccines ?

@DhanyaHerath mentioned that she did see a requirement in pacific countries where a facility would prefer to use different pack size as the base unit. Example: warehouse deals with bottles, dispensary deals with ml, to achieve this we have to 'force' warehouse to deal in ml.

Can use fraction as the pack size multiplier for existing customer (i.e. if they deal with vials already, but we want to deal in doses, a doses unit could have 0.1 multiplier for vial with 10 doses)

Would facility deal with just one pack size or multiple pack sizes (multiple unit types ?) Like would a facility want to deal in vials and in doses at the same time ? Wouldn't this create the original confusion (but instead of packs we can deal with unit type) ? Does it mean we have to deal with different unit types in different UI areas ? Or have a way to `mass` select unit type ?

For the above, it was suggested maybe we need to allow for not just default unit type for facility, but being able to select unit type for an item (as in the design should allow to transition to that) ?
 
It does sound like quite a bit of configuration would be needed, potentially. On configuration front, it's a good idea to `label` unit type (maybe short and long name), rather then just using the pack size multiplier. Basically just like:

nested_units
Except we don't really need to 'nest' them, we just say what the multiplier of the base unit should be.

There a bit more logic for vaccines that would most likely require a different solution (more vaccine targeted):
* An 'extra' dose sometimes sneaks in ¯\_(ツ)_/¯ (bonus dose), see top right of screenshot on this page under [Step four: Dispense a vaccine](https://docs.msupply.foundation/mobile/dispensing_vaccines/vaccine-dispensing), unfortunately I could only find [this issue](https://github.com/openmsupply/mobile/issues/3921) as source of requirement
* Unlike most other medicines, once vial is open it will not last very long (day, two ?) and would be 'wasted' (open vial wastage), open vial wastage should really be recorded in doses, how should this be captured ? (vaccine management design needs to account for this, I don't think it's simply solved by displaying doses rather then vials in our UI).

**Comment by GaryWilletts** [comment]
I have all sorts of alarm bells going off with this one! I honestly think option 1 (clear UI and training) is the way to go. I can see Option 2 being irritatingly complicated for users (and does it impy that when you change pack size you might also have to change the units?). Vials and doses are well handled in existing mSupply so you can get the solution for that from there. I've never seen users wanting to split bottles at lower level facilities. I think the only improvements we need at the moment are clarity in the UI.

One other thing - don't underestimate the impact of the decision that is made regarding this KDD!

**Comment by richardmoizeau** [comment]
Sorry that I'm not gonna try to find the best option but from the experience that I had in Djibouti and discussions I had with stakeholders: 

- in Djibouti, the main issue was that they deal with packs IRL. But it's all good in Open mSupply cause you can issue number of packs (O/S), you receive packs (I/S), you count packs (S/T) except for orders, cause you order in units... So they asked us to change the UI so they could enter an order quantity in qty of "default" pack sizes and then if they could authorise a number of default pack sizes in the remote authorisation module.  That works well for Global Fund products cause their products are very "stable" and always come in the same pack sizes... (almost always...)

- my take working in a warehouse staff is that warehouse only cares about SKU and the SKU is what they buy and distribute so most of the times packs (on packing slip, you usually see a number of packs, not units) and you usually have more than one code for different pack size (Item A `Paracetamol 500mg 100 Tabs`, Item B `Paracetamol 500mg 20 tabs`, etc...). When a warehouse issue a pack of item A, he does not invoice for 100 tabs, he invoice for 1 pack. 

### **for vaccines, it starts to be tricky IMHO** 
- cause the global allocation (at UNICEF/GAVI level), is done is doses (eg: I give 1 millions doses to Fiji), logical if their target is to vaccinate 500k people 2 times. 
- They have 2 suppliers, one can deliver Item A (packs of 10 vials (10 doses) and the distribute item B, packs of 100 vials (10 doses). When placing the order, warehouse won't order 1 millions doses but: 5000 packs of Item A and 500 packs of Item B
- central warehouse receive packs, store packs and most likely distribute packs. They can repack but they will repack vials in small packs, not doses. So would be logical for warehouse to see quantity of packs and quantity of vials
- regional warehouse may start to be interested in vials and doses cause that's how the demand from their customer is expressed into (eg: I want 300 doses for the 300 new born babies in my city)
- vaccination centres want to see doses cause that what they distribute (or administer) but at the end of the day, they are going to count vials for the stocktakes (don't ask the user to do calculations by himself, his first reactions will be why is the system not doing it). And the open vial wastage metric needs the concept of vial to exist. 
- for reporting, GAVI/UNICEF wants to see how many doses are being stored at all level of the Supply Chain, how doses distributed, how many doses wasted...

In a nutshell:
- central warehouse: packs and vials (1 packs of 100 vials)
- intermediate and health centers: vials and doses (100 vials of 10 doses) but don't forget that even service delivery points receives packs!, hard to carry vials around without a box)

**Comment by Gwalior-84** [comment]
This is a super important issue - thanks for bringing it up @adamdewey and working on the KDD @andreievg .  I agree with @GaryWilletts that we really don't want to get it wrong, so should tread carefully.

There is reference to a meeting...  I think for big issues like this, there needs to be more effort to get wider input from our own people and we should try to get some wise external input - I know one or two in PNG who have strong views on this, and I'm sure we know of consultants who can provide valuable input.

There is also a [Nested units (Tertiary > Secondary > Primary > Base units) Trello card](https://trello.com/c/7hRXZVIh/78-nested-units-tertiary-secondary-primary-base-units) where this issue has been discussed.

**Comment by Chris-Petty** [comment]
> It does sound like quite a bit of configuration would be needed, potentially. On configuration front, it's a good idea to `label` unit type (maybe short and long name), rather then just using the pack size multiplier. Basically just like:
> 
> nested_units Except we don't really need to 'nest' them, we just say what the multiplier of the base unit should be.

A question that comes up for me is what exactly this table is stored against? My concern is that a warehouse may procure medicines from multiple suppliers, where all the nesting levels are different. Without defining multiple duplicate items, potentially one for each manufacturer, my thoughts are either:

- This table is against item_lines (and maybe trans_lines? Need to think through transfers). Would be a looooooooot of duplication everywhere though.
- There is another table something like `item_variant` that is central data and can be associated with the different variations of packing/nesting. It could also have the brand name... maybe it isn't a new table, but `[item]` leveraging something similar to "cross reference" items. Either way, as @mark-prins mentioned similarity with challenges in universal codes, this solution reminds me of it too. https://codes.msupply.foundation/detail/cf00cc3f 

Aside: it's been mentioned but can't stress enough most of what I've seen here needs to be interoperable across the different versions of different syncing apps (and external integrations!) which will likely be very challenging, possibly a line in the sand where all sites need an upgrade before syncing situation. Haven't thought it through very deeply yet though. Maybe so long as item_lines still have their `number_of_packs` and `pack_size`, interoperability won't be too hard. Just need to be sure!

**Comment by andreievg** [comment]
Thanks for input so far.
I've seen a couple of ppl mention their concern about sync and compatibility etc.. It's well noted, atm none of the options do anything but change the display (as you would see mention of translation at graphql layer).

Interesting enough pack to one preference is the variation of this (and could have been solved with option 2, I think in a nicer way, if there was a distinct layer between service and presentation, which we have with graphql mapping). Also pack to one preference is somewhat of a validation that there is/was a real requirement to treat items with a particular pack size at a particular facility (since we did add it in mSupply)

File: decisions/multi-unit-item.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+- *Outcome*: Option 2 (store preference table with store_id as primary key and columns as preferences)
+- *Related Material*: See original [issue](https://github.com/openmsupply/open-msupply/issues/1089)
+
+## Background
+
+We've noticed user's confusion in understanding pack sizes. This is especially prominent when users transition from existing systems where pack size have been concealed (i.e. multiple variants of the same item, each defining it's own pack size).
+
+It seems like the core requirement is:
+
+* We want to treat an item as having one particular pack size in a particular facility (requirement to be confirmed)
```
[the remark this answers is not in the document]
Comment (andreievg) [reply decisions/multi-unit-item.md:15]: Thanks, very good point, was also mentioned by Richard's comment. 7d070c9707e9155a84f163c8b48ccb7f8f939227

File: decisions/multi-unit-item.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+Have a very solid onboarding during training, focused on concept of packs and units (as the first training task). Showing scenarior where it's important to deal with different packs at the same facility. And explaining the glossary terms that are used in the system and how they relate to these concepts.
+
+*Pros:*
+
+- Most flexible (the system was originally design to be this way for a good reason, but of course original target user group maybe quite different to the user group in this KDD)
+
+*Cons:*
+
+- Ideally our UI/UX would not need to rely too much on training (it would be intuative for a range of users)
```
[the remark this answers is not in the document]
Comment (andreievg) [reply decisions/multi-unit-item.md]: Not sure If i understand your comment, correctly. I think most of our UI/UX now is very intuitive btw. Also maybe some things just need some training ? 

Maybe this is one is also the pro ? ¯\_(ツ)_/¯  db0804a70ab5820a1dfbd1662f9d2c5c36bfa104

**Comment by DhanyaHerath** [comment]
### Notes from meeting, 20/06/2023

**Attending:** Andrei, Mark G, Adam, Alain, Gary, Dhanya 
**Objective:** Get clear on the requirements/what the problem is 

Full notes here: https://docs.google.com/document/d/1LuW0XD1Q0tgrM9YxG_62iAawDopjliDiwXuSAXDyldU/edit 
End of meeting below: 

**Summary of use cases that are not easily met by repack functionality in mSupply:** 
- Intuitiveness 
- Vaccines (vials with extra doses etc, ordering vs dispensing in two units) 
- Dealing with liquids 
- Base units being different between facilities (eg. blisters in one facility vs tablets in another facility - which means you have to try break up beyond a base unit).  
      - Note, this actually seems similar to the liquids use case! 

Can these be dealt with training?
* Yeeees, but can be quite difficult. 
Can we do vaccines for supply chain in omSupply without doing this work?
* Yes, but it's not pretty (in current mSupply) 
* Would potentially need a way to deal with the same item having a vial with different doses 
      * MG: Could potentially use reference items

Other 
* MG: our language needs to be consistent and clear throughout the platform (yes!! +1)

**Comment by Gwalior-84** [comment]
Core requirements in the Background section of the KDD needs expansion:

General goal:  Design a system that is simple enough that it is easy for relatively untrained user to use, and yet flexible enough to meet the legitimate needs of a wide range of stakeholders.  Mission impossible? 😨 

It might help to document which stakeholder the requirement is coming from.

**Core requirement of donor agencies** (GF, UNICEF, UNFPA, etc):  All facilities using mSupply _globally_ can report on the number of 'base units' in stock and transacted (receipts, issues, invads) at each 'store' using mSupply.  Therefore, need to be able to define base units that systems are forced to use.

**Core requirement of users / management:**  Be able to:

- Clearly understand what is being described through the words on the screen
- Manage stock in a range of pack sizes / forms below.
- Note that 'manage' includes 'repack' between forms / pack sizes.

**Pack sizes / forms**
- For tablets / capsules:
  - Base unit (tablet) - normally the minimum dispensable unit
  - Strips
  - Courses (e.g. adult does Artemether + Lumefantrine Tab., 20+120mg (a common anti-malarial treatment) consists of four strips of six tablets).  This is certainly linked to regimens.
  - Boxes
  - Cartons (containing boxes)
  - Pallets
- For liquids
  - ml, normally the minimum dispensable unit
  - bottles
  - cartons
  - pallets
- For vaccines
  - Doses
  - Vials, containing doses, of different number per vial, even for the same item
  - Boxes
  - Cartons
  - Pallets

**Comment by andreievg** [comment]
Force merging this pull request. It's been in discussion for more then a month with consensus reached for option 3
